inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,154 @@
+            return 100000
+        return int(string_time_planner)
+
+    def page(self, time_planner):
+        """ General main method called for GET and POST """
+        username = self.user_manager.session_username()
+        all_courses = self.course_factory.get_all_courses()
+        time_planner = self.time_planner_conversion(time_planner)
+
+        # Get the courses id
+        open_courses = {courseid: course for courseid, course in all_courses.items()
+                        if self.user_manager.course_is_open_to_user(course, username, False) and
+                        self.user_manager.course_is_user_registered(course, username)}
+        open_courses = OrderedDict(sorted(iter(open_courses.items()), key=lambda x: x[1].get_name(self.user_manager.session_language())))
+
+        courses = {self.course_factory.get_course(courseid): course for courseid, course in all_courses.items()
+                   if self.user_manager.course_is_open_to_user(course, username, False) and
+                   self.user_manager.course_is_user_registered(course, username)}
+
+        # Get last submissions for left panel
+        last_submissions = self.submission_manager.get_user_last_submissions(5, {"courseid": {"$in": list(open_courses.keys())}})
+        except_free_last_submissions = []
+        for submission in last_submissions:
+            try:
+                submission["task"] = open_courses[submission['courseid']].get_task(submission['taskid'])
+                except_free_last_submissions.append(submission)
+            except:
+                pass
+
+        # Get the courses tasks, remove finished ones and courses that have no available unfinished tasks with upcoming deadline in range
+        tasks_data = {}
+        outdated_tasks= {}
+        for course in courses:
+            courseid = course.get_id()
+            outdated_tasks[courseid] = []
+            tasks_data[courseid] = {}
+            tasks = course.get_tasks()
+            for task in tasks:
+                the_task = course.get_task(task)
+                if (not the_task.get_accessible_time().is_open()) or ((the_task.get_accessible_time().get_soft_end_date()) > (datetime.now()+timedelta(days=time_planner))):
+                    outdated_tasks[courseid] += [task]
+            new_user_task_list = course.get_task_dispenser().get_user_task_list([username])[username]
+            tasks_data[courseid].update({taskid: {"succeeded": False, "grade": 0.0} for taskid in new_user_task_list})
+            user_tasks = self.database.user_tasks.find({"username": username, "courseid": course.get_id(), "taskid": {"$in": new_user_task_list}})
+            for user_task in user_tasks:
+                tasks_data[courseid][user_task["taskid"]]["succeeded"] = user_task["succeeded"]
+                tasks_data[courseid][user_task["taskid"]]["grade"] = user_task["grade"]
+                if tasks_data[courseid][user_task["taskid"]]["succeeded"]:
```

Comment:
If you put this check right after the for statement, you won't need to upgrade the content of a deleted item.

Reply:
Indeed.